### ISSUE TITLE
Implement named helper variable references

### ISSUE THREAD

**Issue by kpdecker** [wycats/handlebars.js#907]
Implement the remainder of the feature started in #906.

We want to allow for helpers to provide named values that may be accessed by any child program run. The goal is to allow explicit naming of helper data fields so they can easily be accessed in nested scope, vs. arbitrary `@../index` references.

For example:

```
{{#each foos as |foo indexFoo|}}
  {{#each bars as |bar indexBar|}}
    {{indexFoo}}.{{indexBar}}. {{foo}} {{bar}}
  {{/each}}
{{/each}}
```

**Comment by kpdecker** [comment]
Released in 3.0.0

**Comment by ErisDS** [comment]
Could I please trouble someone for a little help understanding how this works? I've been trawling through the code changes, tests, and trying it out, and although I've got it working, I'm not sure on a few things. I'll explain what I understand so far, and hopefully this'll end up being a sort of documentation :grin: 

Taking the first block param test as an example:

```
it('should take presedence over context values', function() {
    var hash = {value: 'foo'};
    var helpers = {
      goodbyes: function(options) {
        equals(options.fn.blockParams, 1);
        return options.fn({value: 'bar'}, {blockParams: [1, 2]});
      }
    };
    shouldCompileTo('{{#goodbyes as |value|}}{{value}}{{/goodbyes}}{{value}}', [hash, helpers], '1foo');
});
```

Inside a block helper, if a block param was specified when the helper was called, we now get access to `options.fn.blockParams` which tells us how many block params the user asked for. In this case it was just 1, which was declared as `value`.

When we call `fn`, we pass the context as the first argument, and can now pass a `blockParams` option as a property on the second (`options`?) argument. `blockParams` should be an array.

The first item of the array is the data we want to resolve for our blockParam. In the the test, the `value` block param is going to resolve to `1`. The test shows that `1` gets output instead of `bar` which is the value of `value` in the context, asserting that the block param takes precedence.

What I don't understand, is what the second item of the array `2` is for? Is this purely for testing purposes to show that passing extra block params doesn't do anything, or does it mean something else?

The `#each` implementation passes an array with 2 properties, essentially the value and the index as you'd expect. It does this regardless of whether 1 or 2 block params are declared when calling the helper. I guess it's simpler to always pass both rather than checking to see what was declared.

The `#each` helper also [adds a `path` property](https://github.com/wycats/handlebars.js/commit/396795c983273bb5ca4dc67ddc74eb12f00bf110#diff-ae3504422051de557ee42ced88f17a34R82) to the `blockParams` array, which so far for me always evaluates to `[ NaN, null ]`. I'm not at all sure what this path property is for, or when it comes into play - it seems to be related to `contextPath` but I'm no clear on what that is either! 

Basically, I'm trying to figure out exactly what's needed and why in order to determine what might be necessary in a custom helper. Thanks in advance!

**Comment by kpdecker** [comment]
@ErisDS we need to do a better job of documenting this for implementors of helpers. 

Your analysis is pretty much spot on, and it might be a good thing for us to pull this into the formal docs.

Yes, `2` is strictly for testing. Wanted to make sure that it didn't blow up if additional values are passed and didn't want the extra time cost of another test for that case explicitly.

Context path is a little used feature similar to the strings mode used by Ember. This is a little used feature that allows code to attempt to lookup a data field from the root context value based on handlebar's iteration. I believe the only consumer of this is Thorax's SSJS implementation and you can likely ignore if you aren't touching that.

**Comment by ErisDS** [comment]
> Context path is a little used feature similar to the strings mode used by Ember. This is a little used feature that allows code to attempt to lookup a data field from the root context value based on handlebar's iteration. I believe the only consumer of this is Thorax's SSJS implementation and you can likely ignore if you aren't touching that.

Do you have a reference link to any sort of example of this in action that I can take a peek at? If I manage to understand it I'll have a go at writing up an in depth explanation.  

Just an FYI - I wrote some documentation on Handlebars for Ghost theme developers: http://themes.ghost.org/docs/handlebars not sure if it's helpful but please steal from it if it is.

**Comment by kpdecker** [comment]
@ErisDS the built in helpers implement context path tracking and there is test coverage. https://github.com/wycats/handlebars.js/blob/master/spec/track-ids.js#L140. The only existing documentation is https://github.com/wycats/handlebars-site/blob/master/src/pages/reference.haml#L338.

**Comment by kpdecker** [comment]
@ErisDS wycats/handlebars-site@33feeb9 has an attempt at this. I always feel like my documentation skills are lacking so any input is appreciated. I actively chose to ignore the context path behaviors in the docs as I think it's a very niche use case that will only confuse further in the basic docs. If we do provide better docs, then it's probably better to document it in a section specific to that.

**Comment by Bertrand** [comment]
@kpdecker I'm currently integrating this in handlebars-objc, and from what I understand from JS implementation, context provided by the helper can be different from the first block parameter. 
I find this very confusing and I cannot find any useful use case. 

Did you have anything specific in mind?

**Comment by Bertrand** [comment]
@kpdecker Oh,maybe you wanted block helpers to be able to set named block params without changing current context. Is that it?

**Comment by kpdecker** [comment]
Context and block helpers are completely separate. That they match in `each` is just a function of that specific helper.

**Comment by Bertrand** [comment]
@kpdecker yep, that is what I figured out. I was just trying to find an example of a helper where this was both useful and not totally confusing for users.
